src/kixi/event_replay.clj

```diff
@@ -0,0 +1,176 @@
+  (->> [(t/year hour)       
+        (t/month hour)
+        (t/day hour)
+        (t/hour hour)]
+       (map str)
+       (map #(if (= 1 (count %))
+               (str "0" %)
+               %))
+       (interpose "/")
+       (apply str)))
+
+(def max-objects 20)
+
+(defn hour->s3-object-summaries
+  ([hour]
+   (hour->s3-object-summaries (:s3-base-dir @config) 
+                              (hour->s3-prefix hour)
+                              nil))
+  ([s3-base-dir prefix marker]
+   (let [list-objects-res (s3/list-objects (merge {:bucket-name s3-base-dir
+                                                   :prefix prefix
+                                                   :max-keys max-objects}
+                                                  (when marker
+                                                    {:marker marker})))]
+     (concat (:object-summaries list-objects-res)
+             (when (:next-marker list-objects-res)
+               (hour->s3-object-summaries s3-base-dir prefix (:next-marker list-objects-res)))))))
+
+(defn object-summary->input-stream
+  [s3-object-summary]
+  (let [{:keys [s3-base-dir
+                target-stream]} @config
+        s3-object (s3/get-object :bucket-name s3-base-dir
+                                 :key (:key s3-object-summary))]
+    (:object-content s3-object)))
+
+(defn uuid
+  [_]
+  (str (java.util.UUID/randomUUID)))
+
+(def event-type-version->partition-key-fn
+  {[:kixi.datastore.file-metadata/updated "1.0.0"] #(get-in % [:kixi.comms.event/payload :kixi.datastore.metadatastore/id])
+   [:kixi.datastore.filestore/upload-link-created "1.0.0"] #(get-in % [:kixi.comms.event/payload :kixi.datastore.filestore//id])
+   [:kixi.heimdall/user-logged-in "1.0.0"] uuid
+   [:kixi.datastore.file/created "1.0.0"] #(get-in % [:kixi.comms.event/payload :kixi.datastore.filestore//id])})
+
+(def event->event-type-version
+  (juxt #(or (:kixi.event/type %) (:kixi.comms.event/key %))
+        #(or (:kixi.event/version %) (:kixi.comms.event/version %))))
+
+(defn event->partition-key-event
+  [event]
+  (if-let [partition-key-fn (->> event
+                                 event->event-type-version
+                                 event-type-version->partition-key-fn)]
+    [(partition-key-fn event)
+     event]
+    (prn "Unknown event type: " (event->event-type-version event))))
+
+(defn send-event
+  [[partition-key event]]
+  :done)
+
+(defn prn-t
+  [x]
+  (prn x)
+  x)
+
+(def execute-replay
+  (comp (mapcat hour->s3-object-summaries)
+        (take 2)
+        (map object-summary->input-stream)
+        input-stream->events
+        (map prn-t)
+        (keep event->partition-key-event)
+        (map send-event)))
+
+(defn aggregate-events
+  ([] 0)
+  ([acc] acc)
+  ([acc x] (inc acc)))
+
+(defn execute
+  []
+  (transduce execute-replay 
+             aggregate-events
+             (hour-sequence)))
+
+(defn exit
+  [status msg]
+  (println msg)
+  (System/exit status))
+
+(defn -main
```

Comment:
always via the config and not dates passed in?

Reply:
This is going to be executed via docker, so preferring env cars, as makes the docker file cleaner.